### PR TITLE
Chapter 12 reset password

### PR DESCRIPTION
**Pull request by Arizar98**

### CODE REVIEW

**Comment by Arizar98** [comment]
Ready

File: app/assets/stylesheets/password_resets.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the PasswordResets controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/password_resets.scss]: remove file that never used

File: app/controllers/account_activations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+      log_in user
+      flash[:success] = "Account activated!"
+      redirect_to user
+    else
+      flash[:danger] = "Invalid activation link"
+      redirect_to root_url
+    end
+  end
+
+end
```
Comment (saobv-2223) [inline app/controllers/account_activations_controller.rb]:
```suggestion
  end
end
```

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: [:edit, :update]
+  before_action :valid_user, only: [:edit, :update]
+  before_action :check_expiration, only: [:edit, :update]
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  before_action :get_user, :check_expiration, :valid_user, only: %i(edit update)
```

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    else
+      flash.now[:danger] = t "email_not_found"
+      render :new
+    end
+  end
+
+  def edit; end
+
+  def update
+    if user_params[:password].empty?
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    if user_params[:password].blank?
```

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  # Checks expiration of reset token.
+  def check_expiration
+    if @user.password_reset_expired?
+      flash[:danger] = t "password_reset_expired"
+      redirect_to new_password_reset_url
+    end
+  end
+
+end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  end
end
```

File: app/helpers/password_resets_helper.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/password_resets_helper.rb]: remove file that never used

File: config/environments/development.rb [hunk 7]
```diff
@@ -1,62 +1,14 @@
...
+  config.action_mailer.raise_delivery_errors = true
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.default_url_options = {host: "localhost:3000" }
+
+  config.action_mailer.smtp_settings = {
+  :user_name => '00618017908729',
+  :password => '1ef4c419cc40ec',
+  :address => 'smtp.mailtrap.io',
+  :domain => 'smtp.mailtrap.io',
+  :port => '2525',
```
Comment (saobv-2223) [inline config/environments/development.rb]: using gem figaro to manage env variables

**Review by saobv-2223 — COMMENTED** [review]
Your PR is having 2 commits, merge to 1 commit. please

**Comment by Arizar98** [comment]
The second commit is due to conflicts from git repo, and I am not sure how to solve those conflicts using local repo. Everytime I pull and resolve, git repo still says that the conflict is still present.

**Comment by Arizar98** [comment]
Ready

**Review by saobv-2223 — COMMENTED** [review]
Your PR is having 2 commits, merge to 1 commit please

**Comment by Arizar98** [comment]
Ready

**Comment by Arizar98** [comment]
Ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**